### PR TITLE
[AutoDiff upstream] Add SIL derivative function type calculation.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add `SILFunctionType::getAutoDiffDerivativeFunctionType`.

It computes the derivative `SILFunctionType` for an "original" `SILFunctionType`, given:

- Differentiability parameter indices
- Differentiability result index
- Derivative function kind
- Derivative function generic signature (optional)
- Other auxiliary parameters

Add doc comments explaining typing rules, preconditions, and other details.

Partially resolves TF-1124.
Unblocks upstreaming other SIL differentiable programming infrastructure.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@rxwei: could you please help suggest additional reviewers, if you feel it's appropriate?

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
@jckarter: would you like to review this PR, or suggest a reviewer? I'm asking you since you seem like a SIL code owner.

This PR is part of our effort to [upstream differentiable programming](https://bugs.swift.org/browse/TF-820). It adds a utility function and should be relatively non-controversial.

If there's no activity, I'll merge within a day to unblock progress. Happy to address feedback at any time!

**Comment by jckarter** [comment]
Makes sense to me.

**Pull request merged**
